jina/types/arrays/search_ops.py

```diff
@@ -0,0 +1,78 @@
+import re
+from typing import Union, Dict
+
+import numpy as np
+
+# if False:
+from .document import DocumentArray
+
+
+class DocumentArraySearchOpsMixin:
+    """ A mixin that provides search functionality to DocumentArrays"""
+
+    operators = ['<', '>', '==', '!=', '<=', '>=', 'any', 'all']
+
+    def find(
+        self,
+        regexes: Dict,
+        traversal_paths: list = ['r'],
+        operator: str = 'all',
+        value: int = 1,
+    ) -> None:
+        """
+        Find Documents that match the regular expressions in `regexes`.
+        If `regexes` contain several regular expressions an `operator` can be used to
+        specify a decision depending on the number `value` of regular expression matches.
+
+        Example: If `len(regexes)=3`,  `value=2` and `operator='=='` then the documents
+                 from the DocumentArray will be accepted if they match exactly 2 regular expressions.
+
+        Example: If `len(regexes)=3`,  `value=2` and `operator='>='` then the documents
+                 from the DocumentArray will be accepted if they match at least 2 regular expressions.
+
+        :param regexes: Dictionary of the form {tag:regex}
+        :param traversal_paths: List specifying traversal paths
+        :param operator: Operator used to accept/reject a document
+        :param value: Number of regex that should match the operator to accept a Document.
+                      This is ignored with operators `any` and `all`.
+        :return: DocumentArray with Documents that match the regexes
+        """
+
+        assert (
+            operator in self.operators
+        ), f'operator={operator} is not a valid operator from {self.operators}'
+        iterdocs = self.traverse_flat(traversal_paths)
+        matched_counts = np.zeros(len(self), dtype=np.int32)
+        filtered = DocumentArray()
+
+        for tag_name, regex in regexes.items():
+            regexes[tag_name] = re.compile(regex)
+
+        for pos, doc in enumerate(iterdocs):
+            for tag_name, pattern in regexes.items():
+                tag_value = doc.tags.get(tag_name, None)
+                if tag_value:
+                    if pattern.match(tag_value):
+                        matched_counts[pos] += 1
+
+        if operator == '<':
```

Comment:
keep this in a dictionary mapping from `operator_name` to operation function or partialfunc

Reply:
good catch, done